### PR TITLE
Use `write.parquet.compression-{codec,level}`

### PR DESCRIPTION
**Pull request by jonashaag**
I had to change the `metadata_collector` code due to https://github.com/dask/dask/issues/7977.

For the `<not set>` case (no compression specified) the tests currently pass locally but they shouldn't as we never set zstd as the default. Not sure what's going on?

Should we default to gzip which is the default for Iceberg according to [this](https://iceberg.apache.org/docs/latest/configuration/) document, or should we use a more reasonable and modern default like zstd level 3?

### CODE REVIEW

File: pyiceberg/io/pyarrow.py
Comment (syun64) [inline pyiceberg/io/pyarrow.py:1727]: How do you feel about checking if `compression_codec is not None` before adding the compression option to the ParquetWriter kwarg?

I think we can use the meaningful distinction between None and string 'NONE' to mean either that the compression codec was not set (and hence should use the default) versus 'NONE' meaning that we'd like to avoid compression altogether. I think this might help clean up some of the if conditions in **test_parquet_compression**

File: tests/integration/test_writes.py
Comment (syun64) [inline tests/integration/test_writes.py:536]: I wonder if passing these expected values in pytest.mark.parametrize would make the test a lot more easy to follow, by allowing us to avoid using these if statements for certain inputs. Same goes for the compression-level as well

File: pyiceberg/io/pyarrow.py
Comment (syun64) [inline pyiceberg/io/pyarrow.py:1725]: Do we need this handling? Could we just interpret string 'none' value to mean something different from not setting the write.parquet.compression-codec property at all?

I think it would be simpler to just propagate the string value directly to the compression option of the [ParquetWriter](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html):

"compression[str](https://docs.python.org/3/library/stdtypes.html#str) or [dict](https://docs.python.org/3/library/stdtypes.html#dict), default ‘snappy’
Specify the compression codec, either on a general basis or per-column. Valid values: {‘NONE’, ‘SNAPPY’, ‘GZIP’, ‘BROTLI’, ‘LZ4’, ‘ZSTD’}."

File: tests/integration/test_writes.py
Comment (syun64) [inline tests/integration/test_writes.py:534]: Yeah just as you noted in the PR description, this issue is interesting @jonashaag ! I'm not sure at a first glance how this is being set as ZSTD by default.

Since ParquetWriter compression default is 'snappy' and Iceberg's is 'gzip', I think we will have to set the default value when parsing the properties regardless, just like you suggested.

**Comment by syun64** [comment]
@jonashaag thank you for raising the issue and putting this PR together so quickly! We are very excited to group this fix in with the impending 0.6.0 release. I've left some comments on the PR - let's also wait for reviews from @Fokko and others, and hopefully get this merged in soon 😄

File: pyiceberg/io/pyarrow.py
Comment (jonashaag) [reply pyiceberg/io/pyarrow.py:1727]: Sorry I don't understand what you mean

File: pyiceberg/io/pyarrow.py
Comment (jonashaag) [reply pyiceberg/io/pyarrow.py:1725]: `uncompressed` is the official string for no compression I guess, see https://iceberg.apache.org/docs/latest/configuration/#read-properties

**Comment by jonashaag** [comment]
Can you start CI @syun64?

File: pyiceberg/io/pyarrow.py
Comment (syun64) [reply pyiceberg/io/pyarrow.py:1725]: I missed that. Makes sense 👍

**Comment by amogh-jahagirdar** [comment]
> For the <not set> case (no compression specified) the tests currently pass locally but they shouldn't as we never set zstd as the default

The default parquet compression is ZSTD at least for the integration test purpose, since the integration tests are using a Java based REST catalog and when the metadata.json file gets written, we'll persist ZSTD by default for Parquet files https://github.com/apache/iceberg/commit/2e291c2b67a643ffe93e139483df55f3639cc39d

File: pyiceberg/io/pyarrow.py
Comment (syun64) [reply pyiceberg/io/pyarrow.py:1727]: Sorry for the confusion. I was trying to find an alternative to using the string "<not set>" to emulate the case where write.parquet.compression-codec isn't set in the test case below.

File: pyiceberg/catalog/rest.py
Comment (syun64) [inline pyiceberg/catalog/rest.py:453]: I'm realizing that in the java code base, **write.parquet.compression-codec** is a [table property](https://github.com/apache/iceberg/blob/fb02bd2d713c9bef8b8ccee334f6d65087c344b2/core/src/main/java/org/apache/iceberg/TableProperties.java#L144C30-L144C49), that can be overridden as a [runtime option](https://iceberg.apache.org/docs/latest/spark-configuration/#write-options) on write.

Does it make sense to keep this logic to copy these properties from the catalog properties, or should we simply rely on the users to [set these as table properties](https://github.com/apache/iceberg-python/blob/e385e7757cac08380a7578cf5a230bd62e88bb36/pyiceberg/catalog/rest.py#L448) on table creation?

Comment (syun64) [reply pyiceberg/catalog/rest.py:453]: For example, like: 

```
tbl = catalog.create_table(identifier=identifier, schema=TABLE_SCHEMA, properties={'format-version': '1', 'write.parquet.compression-codec': compression_codec, 'write.parquet.compression-level' compression_level})
```

File: pyiceberg/io/pyarrow.py
Comment (jonashaag) [reply pyiceberg/io/pyarrow.py:1727]: Sorry I still don't understand what you want me to do

Comment (syun64) [reply pyiceberg/io/pyarrow.py:1728]: Let's resolve this review comment. Your review recent commit fixed the pattern in the test, and we shouldn't expect to see **None** table property values here.

File: tests/integration/test_writes.py
Comment (syun64) [inline tests/integration/test_writes.py:528]:
```suggestion
    tbl = catalog.create_table(identifier=identifier, schema=TABLE_SCHEMA, properties={'format-version': '1', **compression_properties})
```

File: tests/integration/test_writes.py
Comment (syun64) [inline tests/integration/test_writes.py:520]:
```suggestion
            "s3.secret-access-key": "password",
        },
```

File: pyiceberg/catalog/rest.py
Comment (syun64) [inline pyiceberg/catalog/rest.py:456]:
```suggestion
```

**Review by syun64 — COMMENTED** [review]
Suggestions in line with [review comment](https://github.com/apache/iceberg-python/pull/358/files#r1477128027) to use write.parquet.compression-codec as table property instead of catalog property

**Comment by jonashaag** [comment]
I've changed the properties to be table properties and added handling for some other Parquet properites

File: pyiceberg/io/pyarrow.py
Comment (Fokko) [inline pyiceberg/io/pyarrow.py:1765]: Nit:
```suggestion
        if value := table_properties.get(key):
            try:
                return int(value)
            except ValueError as e:
                raise ValueError(f"Could not parse table property {key} to an integer: {value}") from e
        else:
            return int(value)
```

File: pyiceberg/io/pyarrow.py
Comment (Fokko) [inline pyiceberg/io/pyarrow.py:1767]: Do we want to blow up if one of the properties isn't set?

**Review by Fokko — COMMENTED** [review]
This looks great @jonashaag Thanks for jumping on this right away. Once this is in we can do another release candidate 👍 

Could you add this to the docs? This is in `mkdocs/docs/`. I don't think we have a table for this today, but I think we should add it under the paragraph of the file-io: https://py.iceberg.apache.org/configuration/#fileio

File: pyiceberg/io/pyarrow.py
Comment (jonashaag) [reply pyiceberg/io/pyarrow.py:1765]: Ooh the walrus operator already exists in Python 3.8 \o/

File: pyiceberg/io/pyarrow.py
Comment (jonashaag) [reply pyiceberg/io/pyarrow.py:1769]: We want to raise if one of the properties is set. But I guess we should check for None

**Comment by jonashaag** [comment]
Sorry I don't feel comfortable writing documentation because I still lack a lot of Iceberg understanding and terminology. Could you do that part please?

**Comment by Fokko** [comment]
> Sorry I don't feel comfortable writing documentation because I still lack a lot of Iceberg understanding and terminology. Could you do that part please?

Sure thing, no problem at all 👍 

> Should we default to gzip which is the default for Iceberg according to [this](https://iceberg.apache.org/docs/latest/configuration/) document, or should we use a more reasonable and modern default like zstd level 3?

This is a bug in the documentation. See https://github.com/apache/iceberg-docs/pull/305/. Since Java 1.4 it has changed to zstd. But that fix in the documentation has not yet been released (will be part of 1.5).

File: tests/integration/test_writes.py
Comment (Fokko) [reply tests/integration/test_writes.py]: This is a bug in the documentation :( See https://github.com/apache/iceberg-docs/pull/305. Since Java 1.4 it has changed to zstd. But that fix in the documentation has not yet been released (will be part of 1.5).

File: pyiceberg/io/pyarrow.py
Comment (jonashaag) [reply pyiceberg/io/pyarrow.py:1769]: None is not allowed, reverted my changes

**Comment by jonashaag** [comment]
Sweet, ready to merge from my POV

File: pyiceberg/io/pyarrow.py
Comment (Fokko) [inline pyiceberg/io/pyarrow.py:1752]: I checked this through the debugger, and this looks good. Nice change @jonashaag 👍

**Review by Fokko — APPROVED** [review]
Looks good. Thanks for picking this up on such a short notice @jonashaag

File: pyiceberg/io/pyarrow.py
Comment (jonashaag) [reply pyiceberg/io/pyarrow.py:1752]: You can also tell from the PyArrow code that it's identical :)

File: pyiceberg/io/pyarrow.py
Comment (HonahX) [inline pyiceberg/io/pyarrow.py:1778]:
```suggestion
    compression_codec = table_properties.get("write.parquet.compression-codec", "zstd")
```
How about adding the default value here? RestCatalog backend and HiveCatalog explicitly set the default codec at catalog level. https://github.com/apache/iceberg-python/blob/02e64300aee376a76c175be253a29dcd7c31f0cc/pyiceberg/catalog/hive.py#L158

But other catalogs, such as `glue` and `sql`, do not set this explicitly when creating new tables. In general, for tables that have no `write.parquet.compression-codec` key in its property, we still want to use the default codec `zstd` when writing parquet.

**Review by HonahX — COMMENTED** [review]
Thanks for the great work! @jonashaag. Overall LGTM! Just have one comment about the default value.

**Pull request merged**
